### PR TITLE
Add directional pathing

### PR DESCRIPTION
**Pull request by marcushutchings**
Fixes https://github.com/beyond-all-reason/spring/issues/1398

Includes fixes from https://github.com/beyond-all-reason/spring/pull/1395 which Fixes: https://github.com/beyond-all-reason/spring/issues/1386

This change provides the capability to use directional pathing (HAFPS.) I.e. the result of a pathing query takes the direction of the search into account. So you can path down a hill you wouldn't be able to travel up for example.

This change also adds the capability to use shortest route path (QTPFS.) Normally paths are given according to the fastest route. Now it is possible to have units take the shorest route, which helps them follow the expected path of players better - for example, avoiding dangerous areas.

To use directional pathing:

1. Ensure modrule: `movement.allowDirectionalPathing = true`, it is `true` by default.
2. Use a compatible pathing system. Only HAPFS is compatible at the moment. `system.pathFinderSystem = 0`. it is `0` by default.
3. Enable/disable `allowDirectionalPathing` field on any MoveDefs as you see fit. It is enabled by default.

Note: using an incompatible pathing system (like QTPFS) will disable directional pathing, regardless of other settings.
Note: turning off `allowDirectionalPathing` in modrules will also disable directional pathing, regardless of other settings.

To use shortest route pathing:

1. Ensure modrule: `movement.preferShortestPath= true`, it is `false` by default.
2. Use a compatible pathing system. Only QTPFS is compatible at the moment. `system.pathFinderSystem = 1`. it is `0` by default.
3. Enable/disable `preferShortestPath` field on any MoveDefs as you see fit. It is enabled by default.

Note: using an incompatible pathing system (like HAPFS) will disable shortest route pathing, regardless of other settings.
Note: turning off `preferShortestPath` in modrules will also disable shortest route pathing, regardless of other settings.

Notes:

Using directional pathing has a performance COST on HAPFS.
Using shortest route pathing has a performance BENEFIT on QTPFS.

### CODE REVIEW

**Comment by marcushutchings** [comment]
Sorry, I added the QTPFS stuff because the functional code changes is 3 lines. It was more effort to add the boilerplate code to enable the options!

**Comment by GoogleFrog** [comment]
This looks like it changes a bunch of behaviours by default. Also, precisely which settings are expensive?

**Comment by marcushutchings** [comment]
> If both `allowDirectionalPathing` and `preferShortestPath` are now per-movedef then they shouldn't be modrules anymore. This is because it's confusing when there's two vars under the same name in different namespaces and one only works if the other is in a certain state. If somebody still wants a "global" switch then it's trivially done by putting this snippet in `movedefs.lua`:
> 
> ```lua
> for _, md in pairs(MoveDefs) do
>   md.preferShortestPath = false
>   md.allowDirectionalPathing = false
> end
> ```

That is a good point. I'll remove the modrules and I'll have the moveDefs default to false for both flags. That way current behaviour is preserved by default.

**Review by sprunk — APPROVED** [review]

File: rts/Sim/MoveTypes/MoveDefHandler.cpp [hunk 1]
```diff
@@ -153,6 +154,16 @@ void MoveDefHandler::Init(LuaParser* defsParser)
 }
 
 
+void MoveDefHandler::PostSimInit() {
+	// Pathing system is initialized later and the movedefs may need to adjust their behaviour to align with
+	// capabilities of the pathing system.
+	std::for_each(moveDefs.begin(), moveDefs.end(), [](MoveDef& md){
+		md.allowDirectionalPathing &= pathManager->AllowDirectionalPathing();
```
Comment (sprunk) [inline rts/Sim/MoveTypes/MoveDefHandler.cpp:161]: What happens if I use QTPFS and give a raw move order to tell a unit to move across a downward slope? Or if the only available path is downwards? Does the unit slow down?

Comment (lostsquirrel1) [reply rts/Sim/MoveTypes/MoveDefHandler.cpp:161]: The unit's speed is directional regardless of the pathing setting. This has been the case for at least 8 years.

Comment (sprunk) [reply rts/Sim/MoveTypes/MoveDefHandler.cpp:161]: What about this check removed by this PR?

https://github.com/beyond-all-reason/spring/blob/44f0a4d2376dcda6fd21de512976e1507fe54bf0/rts/Sim/MoveTypes/MoveMath/GroundMoveMath.cpp#L31-L36

image

Comment (lostsquirrel1) [reply rts/Sim/MoveTypes/MoveDefHandler.cpp:161]: Ah, well spotted. Well my 8 years comment stands corrected.

Comment (sprunk) [reply rts/Sim/MoveTypes/MoveDefHandler.cpp:161]: If there is no longer way to make units slow down downhill then it's technically a regression, but it looks like every known game used directional movement anyway so nobody is affected. So I think there is nothing to do for now.

Comment (lostsquirrel1) [reply rts/Sim/MoveTypes/MoveDefHandler.cpp:161]: I'm fine to add that option back in. Maybe it would make for a good 'start-here' type ticket?

Comment (sprunk) [reply rts/Sim/MoveTypes/MoveDefHandler.cpp:161]: I think ideally there would be a good default for movement (and IMO the current fits that description) and, down the line, a way to fully control it. Only having a "downhill slowdown is same as uphill" checkbox is super bland and would be the engine only allowing implementing specific game design.

For example, what if I want a simple but custom slope slowdown formula that goes like this? 
* flat < 10°: 0% speed change
* upward ramp, > 10°, < 40°: -33% slowdown
* upward steep ramp, > 40° (as long as within maxslope): -66% slowdown
* downward ramp, > 10°, < 40°: 0% change for movedef INFANTRY, +25% speedup for movedef VEHICLE
* downward ramp, > 40°: -50% slowdown for movedef INFANTRY (gotta tread carefully), +100% speedup for movedef VEHICLE (rollin')

Right now there doesn't seem to be a nice way for a game to control movement speed like this, and the bool option for the previous non-directional formula wouldn't help much either. The questions would then be:
* what would be a good interface to define such formula?
* how well could pathfinders take any custom formulas into account?

It's a scary task (big, but also a lot of opportunity to silently screw things up) and the default seems decent enough.

**Pull request merged**